File: lib/services/slotService.ts
import prisma from '@/lib/prisma';
import { addDays, format, parseISO, startOfDay, addMinutes, isBefore, isEqual, isAfter, startOfToday } from 'date-fns';

export interface SlotInfo {
  start_time: string;
  end_time: string;
  status: 'open' | 'booked' | 'blocked' | 'locked' | 'past';
  price: number;
  booking_id?: string;
  override_id?: string;
  label: string;
}

export class SlotService {
  async getSlotsForDate(turfId: string, slotDateStr: string): Promise<SlotInfo[]> {
    const slotDate = parseISO(slotDateStr);
    const dayOfWeek = slotDate.getDay(); // 0 is Sunday, 1 is Monday ... 
    // Django uses 0=Monday, 6=Sunday. In JS, 0=Sunday, 1=Monday.
    // Wait, let's map it: JS to Django (JS day - 1, if -1 then 6)
    const djangoDayOfWeek = dayOfWeek === 0 ? 6 : dayOfWeek - 1;

    const turf = await prisma.turf.findUnique({
      where: { id: turfId },
      include: {
        schedules: {
          where: { dayOfWeek: djangoDayOfWeek }
        }
      }
    });

    if (!turf) return [];

    let openTime = new Date();
    openTime.setHours(0, 0, 0, 0);
    let closeTime = new Date();
    closeTime.setHours(0, 0, 0, 0);
    let durationMin = 90;

    const schedule = turf.schedules[0];
    if (schedule) {
      if (schedule.isClosed) return [];
      openTime = schedule.openTime;
      closeTime = schedule.closeTime;
      durationMin = schedule.slotDurationMin;
    }

    // 1. Generate all slot windows for the day
    const slots = this.generateSlotWindows(openTime, closeTime, durationMin, turf.basePrice, slotDateStr);

    // 2. Mark booked slots
    const activeBookings = await prisma.booking.findMany({
      where: {
        turfId,
        date: slotDate,
        status: { not: 'CANCELLED' }
      },
      select: { startTime: true, id: true }
    });
    const bookedMap = new Map(activeBookings.map(b => [format(b.startTime, 'HH:mm'), b.id]));

    // 3. Mark overrides
    const overrides = await prisma.slotOverride.findMany({
      where: { turfId, date: slotDate }
    });

    // 4. Check locks (locks expiring in future)
    const activeLocks = await prisma.slotLock.findMany({
      where: {
        turfId,
        date: slotDate,
        expiresAt: { gt: new Date() }
      }
    });
    const lockMap = new Set(activeLocks.map(l => format(l.startTime, 'HH:mm')));

    const now = new Date();
    const todayStr = format(now, 'yyyy-MM-dd');
    const nowTimeStr = format(now, 'HH:mm');

    for (const slot of slots) {
      // Past slots
      if (slotDateStr === todayStr && slot.start_time < nowTimeStr) {
        slot.status = 'past';
        continue;
      }

      if (bookedMap.has(slot.start_time)) {
        slot.status = 'booked';
        slot.booking_id = bookedMap.get(slot.start_time);
        continue;
      }

      // Check overrides
      const slotStartDt = parseISO(`${slotDateStr}T${slot.start_time}:00`);
      let slotEndDt = parseISO(`${slotDateStr}T${slot.end_time}:00`);
      if (slotEndDt <= slotStartDt) {
        slotEndDt = addDays(slotEndDt, 1);
      }

      let overlappingOverride = null;
      for (const ov of overrides) {
        const ovStartDt = parseISO(`${slotDateStr}T${format(ov.startTime, 'HH:mm')}:00`);
        let ovEndDt = parseISO(`${slotDateStr}T${format(ov.endTime, 'HH:mm')}:00`);
        if (ovEndDt <= ovStartDt) {
          ovEndDt = addDays(ovEndDt, 1);
        }

        if (isBefore(slotStartDt, ovEndDt) && isAfter(slotEndDt, ovStartDt)) {
          overlappingOverride = ov;
          break;
        }
      }

      if (overlappingOverride) {
        if (overlappingOverride.status === 'BLOCKED') {
          slot.status = 'blocked';
          slot.override_id = overlappingOverride.id;
          continue;
        } else if (overlappingOverride.status === 'SPECIAL_PRICE') {
          slot.price = overlappingOverride.price || slot.price;
        }
      }

      // Check lock
      if (lockMap.has(slot.start_time)) {
        slot.status = 'locked';
        continue;
      }
    }

    return slots;
  }

  private generateSlotWindows(openTime: Date, closeTime: Date, durationMin: number, basePrice: number, dateStr: string): SlotInfo[] {
    const slots: SlotInfo[] = [];
    
    // Create actual dates for iteration based on hours/minutes
    let dtOpen = new Date(`${dateStr}T${format(openTime, 'HH:mm')}:00`);
    let dtClose = new Date(`${dateStr}T${format(closeTime, 'HH:mm')}:00`);
    
    if (dtClose <= dtOpen) {
      dtClose = addDays(dtClose, 1);
    }

    let current = dtOpen;
    while (addMinutes(current, durationMin) <= dtClose) {
      const next = addMinutes(current, durationMin);
      slots.push({
        start_time: format(current, 'HH:mm'),
        end_time: format(next, 'HH:mm'),
        status: 'open',
        price: basePrice,
        label: `${format(current, 'hh:mm a')} – ${format(next, 'hh:mm a')}`
      });
      current = next;
    }

    return slots;
  }

  async getAvailabilityForRange(turfId: string, startStr: string, endStr: string) {
    const result: Record<string, SlotInfo[]> = {};
    let current = parseISO(startStr);
    const end = parseISO(endStr);

    while (current <= end) {
      const dateStr = format(current, 'yyyy-MM-dd');
      const slots = await this.getSlotsForDate(turfId, dateStr);
      result[dateStr] = slots;
      current = addDays(current, 1);
    }
    return result;
  }
}

export const slotService = new SlotService();

File: backend/services/slot_service.py
"""
SlotService — computes available slots for a turf on a given date.

Algorithm:
  1. Load OperatingSchedule for the day-of-week
  2. Generate all possible slot windows (start_time, end_time) across the operating day
  3. Load all active Bookings for that turf + date
  4. Load all SlotOverrides for that turf + date
  5. Load all Redis checkout locks
  6. Return a list of SlotInfo objects with status: open | booked | blocked | locked
"""
import logging
from dataclasses import dataclass
from datetime import date, time, datetime, timedelta
from typing import List, Optional

from django.conf import settings

from apps.bookings.models import Booking
from apps.turfs.models import Turf, OperatingSchedule, SlotOverride
from services.booking_service import BookingService, get_redis, _slot_lock_key

logger = logging.getLogger(__name__)


@dataclass
class SlotInfo:
    start_time: time
    end_time: time
    status: str          # "open" | "booked" | "blocked" | "locked" | "past"
    price: int
    booking_id: Optional[int] = None
    override_id: Optional[int] = None

    def to_dict(self) -> dict:
        return {
            "start_time": self.start_time.strftime("%H:%M"),
            "end_time": self.end_time.strftime("%H:%M"),
            "status": self.status,
            "price": self.price,
            "booking_id": self.booking_id,
            "label": f"{self.start_time.strftime('%I:%M %p')} – {self.end_time.strftime('%I:%M %p')}",
        }


class SlotService:
    def get_slots_for_date(self, turf: Turf, slot_date: date) -> List[SlotInfo]:
        """
        Return all slot windows for a turf on a given date, each with its status.
        """
        day_of_week = slot_date.weekday()

        try:
            schedule = turf.schedules.get(day_of_week=day_of_week)
            if schedule.is_closed:
                return []
            open_time = schedule.open_time
            close_time = schedule.close_time
            duration_min = schedule.slot_duration_min
        except OperatingSchedule.DoesNotExist:
            # Default to always open (24/7) with 90-minute slots starting from 12:00 AM (00:00)
            open_time = time(0, 0)
            close_time = time(0, 0)
            duration_min = 90

        # 1. Generate all slot windows for the day
        slots = self._generate_slot_windows(
            open_time, close_time, duration_min, turf.base_price
        )

        # 2. Mark booked slots
        active_bookings = Booking.objects.filter(
            turf=turf, date=slot_date
        ).exclude(status=Booking.CANCELLED).values("start_time", "id")
        booked_map = {b["start_time"]: b["id"] for b in active_bookings}

        # 3. Mark overrides
        overrides = list(SlotOverride.objects.filter(turf=turf, date=slot_date))

        # 4. Check Redis locks
        r = get_redis()

        now = datetime.now().time()

        for slot in slots:
            # Past slots
            if slot_date == date.today() and slot.start_time < now:
                slot.status = "past"
                continue

            if slot.start_time in booked_map:
                slot.status = "booked"
                slot.booking_id = booked_map[slot.start_time]
                continue

            # Check override overlap (blocks or special prices)
            slot_start_dt = datetime.combine(slot_date, slot.start_time)
            slot_end_dt = datetime.combine(slot_date, slot.end_time)
            if slot_end_dt <= slot_start_dt:
                slot_end_dt += timedelta(days=1)

            overlapping_override = None
            for ov in overrides:
                ov_start_dt = datetime.combine(ov.date, ov.start_time)
                ov_end_dt = datetime.combine(ov.date, ov.end_time)
                if ov_end_dt <= ov_start_dt:
                    ov_end_dt += timedelta(days=1)

                if slot_start_dt < ov_end_dt and slot_end_dt > ov_start_dt:
                    overlapping_override = ov
                    break

            if overlapping_override:
                if overlapping_override.status == SlotOverride.BLOCKED:
                    slot.status = "blocked"
                    slot.override_id = overlapping_override.id
                    continue
                elif overlapping_override.status == SlotOverride.SPECIAL_PRICE:
                    slot.price = overlapping_override.price or slot.price

            # Check Redis lock
            lock_key = _slot_lock_key(turf.id, slot_date, slot.start_time)
            if r.exists(lock_key):
                slot.status = "locked"
                continue

            slot.status = "open"

        return slots

    @staticmethod
    def _generate_slot_windows(
        open_time: time, close_time: time, duration_min: int, base_price: int
    ) -> List[SlotInfo]:
        slots = []
        dt_open = datetime.combine(date.today(), open_time)
        dt_close = datetime.combine(date.today(), close_time)
        if dt_close <= dt_open:
            dt_close += timedelta(days=1)
        duration = timedelta(minutes=duration_min)

        current = dt_open
        while current + duration <= dt_close:
            slots.append(
                SlotInfo(
                    start_time=current.time(),
                    end_time=(current + duration).time(),
                    status="open",
                    price=base_price,
                )
            )
            current += duration

        return slots

    def get_availability_for_range(self, turf: Turf, start_date: date, end_date: date) -> dict:
        """Returns {date_str: [SlotInfo]} for a date range (used for calendar month view)."""
        result = {}
        current = start_date
        while current <= end_date:
            slots = self.get_slots_for_date(turf, current)
            result[current.isoformat()] = [s.to_dict() for s in slots]
            current += timedelta(days=1)
        return result
